tools/advisor/advisor/db_config_optimizer.py

```diff
@@ -0,0 +1,414 @@
+            # a 30% change in the option's value per iteration. An addend is
+            # also present (+1 or -1) to handle the cases when the option's
+            # old value was 0 or the final int() conversion suppressed the 30%
+            # change made to the option
+            old_value = float(old_value)
+            mul = 0
+            add = 0
+            if action is Suggestion.Action.increase:
+                if old_value < 0:
+                    mul = 0.7
+                    add = 2
+                else:
+                    mul = 1.3
+                    add = 2
+            elif action is Suggestion.Action.decrease:
+                if old_value < 0:
+                    mul = 1.3
+                    add = -2
+                else:
+                    mul = 0.7
+                    add = -2
+            new_value = int(old_value * mul + add)
+        return new_value
+
+    # TODO(poojam23): try other algorithms for improving the database
+    # configuration
+    # currently, this method picks ONE option at random, and makes
+    # appropriate changes to that option for ALL the column families.
+    @staticmethod
+    def improve_db_config(current_config, guidelines):
+        option = random.choice(list(guidelines.keys()))
+        better_config = {option: {}}
+        for action in guidelines[option]:
+            suggested_values = (
+                guidelines[option][action][ConfigOptimizer.SUGG_VAL]
+            )
+            for col_fam in guidelines[option][action][ConfigOptimizer.SCOPE]:
+                old_value = None
+                if (
+                    option in current_config and
+                    col_fam in current_config[option]
+                ):
+                    old_value = float(current_config[option][col_fam])
+                try:
+                    new_value = ConfigOptimizer.apply_action_on_value(
+                        old_value, action, suggested_values
+                    )
+                    better_config[option][col_fam] = new_value
+                except AssertionError:
+                    print(
+                        'WARNING(ConfigOptimizer): provide suggested_values ' +
+                        'for ' + option
+                    )
+        return better_config
+
+    @staticmethod
+    def improve_db_config_v2(options, rule, suggestions_dict):
+        '''
+        this method takes ONE 'rule' and applies all its suggestions on the
+        appropriate options whereas
+        'improve_db_config' picks ONE option to update and applies all
+        suggestions that were triggered for that option
+        '''
+        # get the current configuration
+        required_options = []
+        rule_suggestions = []
+        for sugg_name in rule.get_suggestions():
+            option = suggestions_dict[sugg_name].option
+            action = suggestions_dict[sugg_name].action
+            if not (option and action):  # suggestion not in required format
```

Comment:
can you explain in the doc what how this situation can happen? does it make sense to print out some debugging information instead of silently swallowing the error?

Reply:
The perform_checks() method of the Suggestion class does the needful when the Rules spec is being parsed by RulesSpec object (rule_parser.py)

Reply:
as discussed offline, if this is not going to be the case, lets replace the if-then with an assert.